### ISSUE TITLE
Rule proposal: No useless `as const` assertion

### ISSUE THREAD

**Issue by isaacparker-at** [typescript-eslint/typescript-eslint#4793]
### Description

My understanding is that `as const` is a no-op when the variable already has a type annotation:

Example:
`const a: {foo: string} = {foo: 'bar'} as const` still results in type `{foo: string}`, since that type takes precedence over `as const`; whereas
`const a = {foo: 'bar'} as const` results in a type of `{readonly foo: 'bar'}`

### Fail

```ts
const a: {foo: string} = {foo: 'bar'} as const // 'as const' doesn't do anything here
```

### Pass

```ts
const a = {foo: 'bar'} as const
```

**Comment by bradzacher** [comment]
Thanks for the request.

I like the idea, but I don't think that this is going to be widespread / common enough to warrant a lint rule.

There is also possibly some cases where the const assertion is necessary (with literal types maybe).

Given the above and also that this case can be [covered by configuring a lint rule](https://typescript-eslint.io/docs/linting/troubleshooting#how-can-i-ban-specific-language-feature), I don't think it belongs in our general-purpose plugin.
